src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Folding/FoldingRangeEndpoint.cs

```diff
@@ -0,0 +1,141 @@
+
+        public FoldingRangeEndpoint(
+            RazorDocumentMappingService documentMappingService,
+            ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+            DocumentResolver documentResolver,
+            ClientNotifierServiceBase languageServer,
+            DocumentVersionCache documentVersionCache)
+        {
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+            _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher ?? throw new ArgumentNullException(nameof(projectSnapshotManagerDispatcher));
+            _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
+            _documentVersionCache = documentVersionCache ?? throw new ArgumentNullException(nameof(documentVersionCache));
+        }
+
+        public FoldingRangeRegistrationOptions GetRegistrationOptions(FoldingRangeCapability capability, ClientCapabilities clientCapabilities)
+            => new()
+            {
+                DocumentSelector = RazorDefaults.Selector,
+            };
+
+        public async Task<Container<FoldingRange>?> Handle(FoldingRangeRequestParam @params, CancellationToken cancellationToken)
+        {
+            var documentAndVersion = await TryGetDocumentSnapshotAndVersionAsync(
+                @params.TextDocument.Uri.GetAbsoluteOrUNCPath(),
+                cancellationToken).ConfigureAwait(false);
+
+            if (documentAndVersion is null)
+            {
+                return EmptyFoldingRange;
+            }
+
+            var (document, version) = documentAndVersion;
+            if (document is null || cancellationToken.IsCancellationRequested)
+            {
+                return EmptyFoldingRange;
+            }
+
+            var codeDocument = await document.GetGeneratedOutputAsync().ConfigureAwait(false);
+            if (codeDocument.IsUnsupported())
+            {
+                return null;
+            }
+
+            var requestParams = new RazorFoldingRangeRequestParam
+            {
+                DocumentHostVersion = version,
+                TextDocument = @params.TextDocument
+            };
+
+            var delegatedRequest = await _languageServer.SendRequestAsync(LanguageServerConstants.RazorFoldingRangeEndpoint, requestParams).ConfigureAwait(false);
+            var foldingResponse = await delegatedRequest.Returning<RazorFoldingRangeResponse?>(cancellationToken).ConfigureAwait(false);
+
+            if (foldingResponse is null)
+            {
+                return EmptyFoldingRange;
+            }
+
+            List<FoldingRange> mappedRanges = new();
+
+            foreach (var foldingRange in foldingResponse.CSharpRanges)
+            {
+                var range = new Range(
```

Comment:
A comment here explaining about the nullable int might be a good idea, particularly if it turns out that its important for non-code-block C#

Also, wouldn't be terrible to put this conversion into a separate method